fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -963,189 +970,249 @@ double MultiVersionDatabase::getMainThreadBusyness() {
 	return 0;
 }
 
-// Returns the protocol version reported by a quorum of coordinators
+// Returns the protocol version reported by the coordinator this client is connected to
 // If an expected version is given, the future won't return until the protocol version is different than expected
 ThreadFuture<ProtocolVersion> MultiVersionDatabase::getServerProtocol(Optional<ProtocolVersion> expectedVersion) {
-	// TODO: send this out through the active database
-	return MultiVersionApi::api->getLocalClient()
-	    ->api->createDatabase(clusterFilePath.c_str())
-	    ->getServerProtocol(expectedVersion);
-}
+	return dbState->versionMonitorDb->getServerProtocol(expectedVersion);
+}
+
+MultiVersionDatabase::DatabaseState::DatabaseState(std::string clusterFilePath, Reference<IDatabase> versionMonitorDb)
+  : clusterFilePath(clusterFilePath), versionMonitorDb(versionMonitorDb),
+    dbVar(new ThreadSafeAsyncVar<Reference<IDatabase>>(Reference<IDatabase>(NULL))) {}
+
+// Adds a client (local or externally loaded) that can be used to connect to the cluster
+void MultiVersionDatabase::DatabaseState::addClient(Reference<ClientInfo> client) {
+	ProtocolVersion baseVersion = client->protocolVersion.normalizedVersion();
+	auto [itr, inserted] = clients.insert({ baseVersion, client });
+	if (!inserted) {
+		// SOMEDAY: prefer client with higher release version if protocol versions are compatible
+		Reference<ClientInfo> keptClient = itr->second;
+		Reference<ClientInfo> discardedClient = client;
+		if (client->canReplace(itr->second)) {
+			std::swap(keptClient, discardedClient);
```

Comment:
does this have an effect?

Reply:
My memory of writing this is a bit hazy, but it looks like what this is saying is that if you add a new client with the same protocol as one we've already added, then it will by default keep the one you've already added. However, if the new one is in some way preferable (as determined by `canReplace`), then it will use the new one instead.